File: ArrayAndHashing/ContainsDuplicate/cd.py
from typing import List


class Solution:

    # NOTE: Simplified further
    def hasDuplicate(self, nums: List[int]) -> bool:
        return len(set(nums)) < len(nums)

    # NOTE: Cleaner approach
    # def hasDuplicate(self, nums: List[int]) -> bool:
    #     seen = set()
    #     for n in nums:
    #         if n in seen:
    #             return True
    #         seen.add(n)
    #     return False

    # NOTE: Original solution
    # def hasDuplicate(self, nums: List[int]) -> bool:
    #     freq = {}
    #     for n in nums:
    #         if n not in freq:
    #             freq[n] = 1
    #         else:
    #             freq[n] += 1
    #
    #     for _, v in freq.items():
    #         if v > 1:
    #             return True
    #
    #     return False


if __name__ == "__main__":
    s = Solution()
    n1 = [1, 2, 3, 3]
    print(s.hasDuplicate(n1))

    n2 = [1, 2, 3, 4]
    print(s.hasDuplicate(n2))

    n3 = []
    print(s.hasDuplicate(n3))

File: ArrayAndHashing/ContainsDuplicate/cd.go
package main

import "fmt"

func hasDuplicate(nums []int) bool {
	set := make(map[int]struct{})

	for _, n := range nums {
		set[n] = struct{}{}
	}

	return len(set) < len(nums)
}

func main() {
	n := []int{1, 2, 3, 3}
	fmt.Printf("T1: %t\n", hasDuplicate(n))

	n2 := []int{1, 2, 3, 4}
	fmt.Printf("T2: %t\n", hasDuplicate(n2))

	n3 := []int{}
	fmt.Printf("T3: %t\n", hasDuplicate(n3))
}
